ASL: desugar case statement (#1131)
herd/herdtools7: asllib/doc/Statements.tex

\Ncaseotherwise \derives \ & \Totherwise \parsesep \Tarrow \parsesep \Nstmtlist &
\end{flalign*}

\hypertarget{def-casediscriminantterm}{}
\hypertarget{def-casealternativeterm}{}
\hypertarget{def-otherwisecaseterm}{}
The expression following the \Tcase\ keyword is called the \emph{\casediscriminantterm}.
The list following the \Tof\ keyword consists of \emph{\casealternativesterm},
optionally ending with an \emph{\otherwisecaseterm}, which follows the \Totherwise\ keyword.

\subsection{Abstract Syntax}

> Case statements are considered syntactic sugar and are desugared-away via \\
> \nameref{sec:ASTRule.DesugarCaseStmt}
> by considering the following requirements:

Reviewer: suggestion:
```suggestion
Case statements are considered syntactic sugar, and are desugared via \\
\nameref{sec:ASTRule.DesugarCaseStmt}, which obeys the following requirements:
```

\RequirementDef{CaseDiscriminant} The \casediscriminantterm\ of a \texttt{case}
statement should be evaluated only once each time the case statement is evaluated.

> \RequirementDef{CaseAlternatives} The \casealternativesterm\ are examined in sequential order.

Reviewer 2: define sequential order please

Author: rephrased

If any of the patterns match the \casediscriminantterm (and the guard

> expression is true, if present --- see below) then this \casealternativeterm\ is considered selected,

Reviewer 2: please replace "see below" with a pointer

Author: removed - I don't think this part of the sentence is needed

  \vepattern \eqdef \EPattern(\vezero, \vpattern)\\
  \vcond \eqdef \choice{\vwhere = \langle\vewhere\rangle}{\EBinop(\BAND, \vepattern, \vewhere)}{\vepattern}
}{
  \casetocond(\vezero, \vcase, \vtail) \astarrow \overname{\SCond(\vcond, \vstmt, \vtail)}{\news}
}
\end{mathpar}

\subsection{Typing}

> Since case statements are transformed into other statements,

Reviewer 2: above the term "desugared-away" is used - here the term "transformed" is used. Please pick one (and define it)

they do not require type system rules.
\lrmcomment{This is related to \identr{WGSY}.}

\subsection{Semantics}

> Since case statements are transformed into other statements,

Reviewer 2: above the term "desugared-away" is used - here the term "transformed" is used. Please pick one (and define it)
